Isometry object

patapufate opened the pull request.

patapufate committed:
first Vector3 Class. Is not separated (h and cc)

patapufate committed:
overload operator <<

patapufate committed:
file h, cc division

patapufate committed:
Cleaning code

patapufate committed:
Vector3 without () overload and cross product

patapufate committed:
Fixing overloading []

patapufate committed:
[] overload and Vector cross method

patapufate committed:
Vector3

patapufate committed:
.gitignore

patapufate committed:
Merge branch 'Vector3Object' of https://github.com/patapufate/cppl1_q12020 into Vector3Object

patapufate committed:
Fixing code review

patapufate committed:
Pull request Fix

patapufate committed:
Pull Request fix

patapufate committed:
Matrix Class

patapufate committed:
Vector3 += -= overloaded operators

patapufate committed:
Vector3 isometric_TEST == OK

patapufate committed:
rm MatrixObject

patapufate committed:
Vector3 Object

Commit:
Matrix3 Object

Commit:
Second Review of Vector3 Object

Commit:
verifying matrix object

Commit:
modifying initialize list

Commit:
Third PR
Vector3 modification from PR
Matrix3 First review

Commit:
Isometry object

Commit:
Isometry Object

glpuga commented on course/.vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,16 @@
+{
+    "configurations": [
```
Please remove the vscode configuration from the repo.

glpuga replied:
This still needs to get done.

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+class Matrix3
+{
+public:
+    explicit Matrix3();
```
No need to use explicit if there are no parameters.

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+    ~Matrix3();
+
+    double det() const;
+    Vector3 &row(const int &index) const;
```
Something's weird if this builds ok. You should not be able to return a non-const reference from a const function.

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+    double det() const;
+    Vector3 &row(const int &index) const;
+    Vector3 &row(const int &index);
+    Vector3 col(const int &index) const;
```
Same.

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+    Vector3 &operator[](const int &index);
+
+    friend bool operator==(const Matrix3 &p, const Matrix3 &q);
+    friend bool operator==(const Matrix3 &p, const std::initializer_list<double> &list);
```
Should also have one for the other way around (initialization list on the left).

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <iostream>
+#include "math.h"
```
Please use cmath instead.

glpuga replied:
and use `<>`

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,62 @@
+#pragma once
+#include <iostream>
+#include "math.h"
```
`<cmath>`

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,62 @@
+    Vector3(const Vector3 &p2);
+    ~Vector3();
+
+    double &x() const { return values[0]; }
```
This should return a `const &`, otherwise you can change the state of the const object.

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,62 @@
+    friend Vector3 operator/(const Vector3 &p, const Vector3 &q);
+
+    friend bool operator==(const Vector3 &p, const Vector3 &q);
+    friend bool operator==(const Vector3 &p, const std::initializer_list<double> &list);
```
Please add another for the other way around.

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,62 @@
+    friend bool operator==(const Vector3 &p, const std::initializer_list<double> &list);
+
+    friend bool operator!=(const Vector3 &p, const Vector3 &q);
+    friend bool operator!=(const Vector3 &p, const std::initializer_list<double> &list);
```
Same.

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,62 @@
+    static const Vector3 kOnes;
+
+private:
+    double *values = new double[3];
```
Please move this into a constructor.

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,55 @@
+    static const Matrix3 kOnes;
+
+private:
+    Vector3 *row_values = new Vector3[3];
```
No need to make this one a pointer to vectors. For the sake of using move operations, it's enough to do it in Vector3.

glpuga commented on course/include/isometry.h:
```diff
@@ -1,17 +1,46 @@
+    Vector3 translation() const;
+    Matrix3 rotation() const;
+
+    friend bool operator==(const Isometry &p, const Isometry &q);
```
Please add `!=`

glpuga commented on course/include/isometry.h:
```diff
@@ -1,17 +1,46 @@
-  explicit Vector3(int value) : value_(value) {}
-  int value() const { return value_; }
+    ~Isometry();
+    static Isometry FromTranslation(const std::initializer_list<double> &list);
```
These functions should be `const` too.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+}
+Matrix3::Matrix3(const std::initializer_list<double> &list)
+{
+    int i = 0, j = 0;
```
Please check the size of the list first.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    row(2) = m.row(2);
+}
+
+Matrix3::~Matrix3() {}
```
I commented to remove the pointer, but if you still wanted to use it you should release the memory here, or you'll be leaking memory.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+double Matrix3::det() const
+{
+    Vector3 i = row(0);
```
This can be a `const & ` and avoid a copy.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+Vector3 &Matrix3::row(const int &index) const
+{
+    try
```
Please check the range of the index.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+}
+Vector3 &Matrix3::row(const int &index)
+{
+    try
```
Please check the range of the index.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+Vector3 Matrix3::col(const int &index) const
+{
+
```
Please check the range of the index.

glpuga replied:
> Please check the range of the index.

It looks like this was unaddressed.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    return (i.x() * j.x() + i.y() * j.y() + i.z() * j.z());
+}
+
+Vector3 &Matrix3::row(const int &index) const
```
Integers are cheap to copy; use a plain `int &`.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+        std::cerr << "Error occurred: " << ex.what() << std::endl;
+    }
+}
+Vector3 &Matrix3::row(const int &index)
```
Same.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    }
+}
+
+Vector3 Matrix3::col(const int &index) const
```
Same.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+        Vector3 p;
+        p.x() = row_values[0][index];
+        p.y() = row_values[1][index];
+        p.z() = row_values[2][index];
```
Please use the x,y,z constructor and do construction in a single step.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    return (Vector3::kZero);
+}
+
+Vector3 Matrix3::col(const int &index)
```
No need to use a reference.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+Vector3 Matrix3::col(const int &index)
+{
+    try
```
Please check the range.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+const Vector3 &Matrix3::operator[](const int &index) const
+{
+    try
```
Please check the range of the index.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+{
+    try
+    {
+        return (row_values[index]);
```
No need for the extra parenthesis here.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+Matrix3 operator+(const Matrix3 &p, const Matrix3 &q)
+{
+    Matrix3 r;
+    r.row(0) = p.row(0) + q.row(0);
```
Can do 
```
Matrix3 r(p);
r += q;
```

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    r.row(0) = p.row(0) + q.row(0);
+    r.row(1) = p.row(1) + q.row(1);
+    r.row(2) = p.row(2) + q.row(2);
+    return (r);
```
No need for parenthesis.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+    r.row(0) = p.row(0) - q.row(0);
+    r.row(1) = p.row(1) - q.row(1);
+    r.row(2) = p.row(2) - q.row(2);
+    return (r);
```
Same, here and below.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+Matrix3 operator-(const Matrix3 &p, const Matrix3 &q)
+{
+    Matrix3 r;
+    r.row(0) = p.row(0) - q.row(0);
```
Same about using constructor and `-=` operator.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+
+bool operator==(const Matrix3 &p, const std::initializer_list<double> &list)
+{
+
```
Please remove the space.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+Matrix3 operator/(const Matrix3 &p, const Matrix3 &q)
+{
+    Matrix3 res;
+    for (size_t i = 0; i < 3; i++)
```
Prefer pre-increment to post-increment, unless you really need it.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+{
+    Matrix3 res;
+    for (size_t i = 0; i < 3; i++)
+        res.row(i) = p.row(i) / q.row(i);
```
Please always use `{ } ` around control blocks.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+bool operator!=(const Matrix3 &p, const std::initializer_list<double> &list)
+{
+    Matrix3 q(list);
+    return (!(p == q));
```
Can be `!(p == list)`.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,236 @@
+std::ostream &operator<<(std::ostream &os, const Matrix3 &p)
+{
+    os << "[";
+    for (size_t i = 0; i < 2; i++)
```
Pre-increment. https://isocpp.org/blog/2016/08/quick-q-efficiency-of-postincrement-v.s.-preincrement-in-cpp

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,168 @@
+    z() = p2.z();
+}
+
+Vector3::~Vector3() {}
```
There's a memory leak if you don't release the pointer.

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,168 @@
+
+Vector3 Vector3::cross(const Vector3 &q) const
+{
+    Vector3 r;
```
Better use the x, y, z constructor to avoid constructing and then rewriting the object.

glpuga reviewed (changes requested):
I only got to about half of it. Good work. Please address these comments and I'll review again. Many of the comments repeat between matrix and vector; please after addressing my comments do a quick self-review looking for issues that may be repeated in different places.

Let me know when this is ready for the next review.

Commit:
Second PR

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,65 @@
+        static const Vector3 kOnes;
+
+    private:
+        double values[3];
```
What happened with the pointer to a vector? This seems to have gone a couple of versions back.

patapufate replied:
Yes. I undestand in the first PR that I have to switch again to vector.
I fixed using smart pointers (Take look, because I don't know if is ok). 

std::shared_ptr<double[]> values;

Because of this I remove The destructor.

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,65 @@
+        Vector3();
+        explicit Vector3(const double x, const double y, const double z);
+        Vector3(const std::initializer_list<double> &list);
+        Vector3(const Vector3 &p2);
```
Please add a move constructor too.

patapufate replied:
Done. Please. Can you take a look?

patapufate replied:
I resolved with shared pointers.

glpuga replied:
> I resolved with shared pointers.

That was not my intention. There was an educational purpose behind using raw pointers for this. Using smart pointers eliminates the problem, but it was a problem that we created only to practice move semantics.

glpuga commented on course/include/Vector3.h:
```diff
@@ -0,0 +1,65 @@
+        double &operator[](const int index);
+
+        Vector3 &operator+=(const Vector3 &q);
+        Vector3 &operator-=(const Vector3 &q);
```
Please also add copy and move assingment operators.

patapufate replied:
Done. Please. Can you take a look?

patapufate replied:
I resolved with shared pointers.

glpuga replied:
Same.

glpuga commented on course/include/isometry.h:
```diff
@@ -1,17 +1,47 @@
-};
+        Vector3 transform(const Vector3 &v) const;
+        Vector3 transform(const std::initializer_list<double> &list) const;
+        Isometry inverse(void) const;
```
nit: no need to use `(void)`. That's really old C.

glpuga commented on course/include/isometry.h:
```diff
@@ -1,17 +1,47 @@
+    private:
+        Vector3 vector3_isometry;
+        Matrix3 matrix3_isometry;
+        double v_row3_isometry[4];
```
hmmmm this fourth row is not really needed.

patapufate replied:
Yes. You're right. I saw some transformation in time that use this, but, in our exercise we don't use it at all. I will remove

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+        if (list.size() == 9)
+        {
+            try
+            {
```
nit: better put the try/catch on the most external block.

patapufate replied:
fixed

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+        int i = 0, j = 0;
+        if (list.size() == 9)
+        {
+            try
```
I don't think that anything within this try/catch block can actually throw.

patapufate replied:
Done

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+
+    Matrix3::Matrix3(const std::initializer_list<double> &r0, const std::initializer_list<double> &r1, const std::initializer_list<double> &r2)
+    {
+        try
```
Nothing within this block can throw.

patapufate replied:
Done

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+        row(2) = m.row(2);
+    }
+
+    Matrix3::~Matrix3() {}
```
Empty destructors can be left implicit.

patapufate replied:
Done

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+
+    const Vector3 Matrix3::row(const int index) const
+    {
+        try
```
You're overusing these a bit; please check other uses of try/catch and check if they are really needed.

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+    {
+        try
+        {
+            if (index < 3)
```
What happens if index is valued `-100`?

patapufate replied:
Fixed

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+    {
+        try
+        {
+            if (index < 3)
```
Check negatives.

patapufate replied:
Fixed

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+
+        try
+        {
+            if (index < 3)
```
Negatives.

patapufate replied:
Done

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+    {
+        try
+        {
+            if (index < 3)
```
Negatives.

patapufate replied:
Fixed

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+    {
+        try
+        {
+            if (index < 3)
```
Negatives.

patapufate replied:
Fixed

glpuga commented on course/src/Matrix3.cc:
```diff
@@ -0,0 +1,286 @@
+    {
+        try
+        {
+            if (index < 3)
```
Negatives.

patapufate replied:
Fixed

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,202 @@
+        if (list.size() == 3)
+        {
+            int i = 0;
+            try
```
No need, nothing throws.

patapufate replied:
Fixed

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,202 @@
+        if (index < 3)
+        {
+
+            try
```
No need.

patapufate replied:
Fixed

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,202 @@
+
+    const double &Vector3::operator[](const int index) const
+    {
+        if (index < 3)
```
Negatives.

patapufate replied:
Done

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,202 @@
+
+    double &Vector3::operator[](const int index)
+    {
+        if (index < 3)
```
Negatives.

patapufate replied:
Fixed

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,202 @@
+    {
+        if (index < 3)
+        {
+            try
```
Check other uses of try/catch for need in this file too.

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        v_row3_isometry[0] = 0;
+        v_row3_isometry[1] = 0;
+        v_row3_isometry[2] = 0;
+        v_row3_isometry[3] = 1;
```
No need for a fourth row. All operations can be done just using the translation and the rotation.

patapufate replied:
Fixed

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+    {
+        if (list.size() == 3)
+        {
+            try
```
Please check try/catch blocks.

patapufate replied:
Fixed

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    Isometry::~Isometry()
```
Empty destructors can be left out.

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+    Isometry Isometry::RotateAround(const Vector3 &v, const double &a)
+    {
+        Isometry I1;
+        if (v == Vector3::kUnitX)
```
This code should be able to rotate around any vector, and not just the canonical three.

glpuga replied:
Look here: https://en.wikipedia.org/wiki/Rotation_matrix#Axis_and_angle

patapufate replied:
Done.

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+
+    Isometry Isometry::inverse(void) const
+    {
+        return Isometry(vector3_isometry * (-1), matrix3_isometry);
```
This is not the inverse transformation. Ping me and we'll talk about it.

patapufate replied:
Done. I Also Included a Test to Isometry Test

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+
+    Isometry Isometry::compose(const Isometry &I1) const
+    {
+        return Isometry((*this) * I1);
```
Can be `return (*this) * I1; `

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        {
+            a.x() = p.row(i).x();
+            a.y() = p.row(i).y();
+            a.z() = p.row(i).z();
```
This means that `a` is `row(i)`?

patapufate replied:
Yes. Fixed

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+                b = q.col(j);
+                m.matrix3_isometry[i][j] = a.dot(b);
+            }
+        }
```
I'm on the idea that this algorithm is not correct, but even if it is, there's a much better (and shorter) way to do it. Ping and we'll talk it over.

patapufate replied:
I moved the Matrix Multiplication to the Matrix Object. I also included a test to Isometry_Test

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        return os;
+    }
+
+    Isometry &Isometry::operator+=(const Isometry &q)
```
Adding or subtracting isometries together is not really a valid operations, since the result is almost never a valid transformation.

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        return *this;
+    }
+
+    Isometry &Isometry::operator-=(const Isometry &q)
```
Same.

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        return *this;
+    }
+
+    Isometry operator+(const Isometry &p, const Isometry &q)
```
Same.

patapufate replied:
Done

glpuga commented on course/src/isometry.cc:
```diff
@@ -0,0 +1,245 @@
+        return Isometry(v, m);
+    }
+
+    Isometry operator-(const Isometry &p, const Isometry &q)
```
Same.

patapufate replied:
Done

glpuga commented on course/test/src/isometry_TEST.cc:
```diff
@@ -58,10 +95,19 @@ GTEST_TEST(Vector3Test, Vector3Operations) {
+  {
+    t[4] = 1;
+  }
+  catch (std::out_of_range const &err)
```
Maybe this is some test that got forgotten in the code?

patapufate replied:
Done

glpuga reviewed (changes requested):
Good work. Still some fixes to do, and I'd like to be copy/move semantics in use in this code. Please address my comments in fixup commits, so that its easier for me to determine what you updated in the code in the corrections.

glpuga commented:
Also, it's usually a good idea to answer to each of the review comments with a short ack such as "Done." or "Fixed." just so that the reviewer knows that you read and addressed a particular comment. It also allows you to make sure that you addressed all of them.

Commit:
Review from second PR

glpuga commented:
Hi @patapufate , is this ready for review? If so, did you push your changes to the repo?

Commit:
Isometry Object Third Review

Isometry Inverse and Matrix inverse

Removed col()

Isometry object

Backup

Isometry Object with smart Pointers

Isometry Object Third Review

glpuga commented on course/include/Matrix3.h:
```diff
@@ -19,7 +19,9 @@ namespace ekumen
         const Vector3 col(const int index) const;
-        Vector3 col(const int index);
+
+        Matrix3 mul(const Matrix3 &m1, const Matrix3 &m2) const;
```
The first operand for this operations would better be the instance on which this member gets called; otherwise you can make the member `static` instead of `const`, so that it becomes a class method with two instances as arguments.

Like this, it needs an instance on which to be called (because it is not static), but if I want to multiply that same instance with another matrix, then I have to explicitly pass it as an argument even though it is already available as `*this`.

patapufate replied:
I don't understand this comment.

patapufate replied:
Fixed. sorry for the last comment

glpuga commented on course/src/Vector3.cc:
```diff
@@ -4,48 +4,43 @@ namespace ekumen
-    { 
-        values = std::shared_ptr<double[]>(new double[3]);        
+    {
+        values = std::shared_ptr<double[]>(new double[3]);
```
Don't use smart pointers here, but when you need to use them, don't use them along with `new`. Use `std::make_shared()`. Please research the advantages of this method.

patapufate replied:
removed smart pointers

glpuga commented on course/src/isometry.cc:
```diff
@@ -160,23 +153,9 @@ namespace ekumen
-        }
-        m.vector3_isometry = i.vector3_isometry + i2.vector3_isometry;
-        return m;
+        Matrix3 m1 = m1.mul(i.matrix3_isometry, i2.matrix3_isometry);
```
Notice here what I meant about the definition of the `mul()` method. Your using the `m1` instance, which is not yet initialized, to be able to run `mul()` on two other matrixes, and then store the result in `m1`. Please, refactor the method so that this works like
```
Matrix3 m1{i.matrix3_isometry.mul(i2.matrix3_isometry)};
```

patapufate replied:
fixed and refactored.

glpuga commented on course/src/isometry.cc:
```diff
@@ -160,23 +153,9 @@ namespace ekumen
-        }
-        m.vector3_isometry = i.vector3_isometry + i2.vector3_isometry;
-        return m;
+        Matrix3 m1 = m1.mul(i.matrix3_isometry, i2.matrix3_isometry);
```
As a side note, it's usually bad practice to use short variable names such as `i` and `i2`. Also, it's usually frown upon to use `i` for anything other than loop indexes, since that's the universally accepted use of `i`, `j` and `k`. Notice that there's people who would object even their use within this narrow context.

[Please read this](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)

glpuga commented on course/include/Matrix3.h:
```diff
@@ -14,11 +14,10 @@ namespace ekumen
-        const Vector3 row(const int index) const;
-        Vector3 &row(const int index);
+        const Vector3 row(int index) const;
+        Vector3 &row(int index);
```
Please restore the  `const` attribute.

patapufate replied:
restored

glpuga reviewed (changes requested):
Some more comments. Please, take a look at [Fabri's solution to move semantics on Vector3](https://github.com/elector102/cppl1_q12020/pull/1/files#diff-dfb4d63ae36c57de208059ac43d21756R82) to have a better picture of what we are aiming for with the move semantics mini-exercise.

Don't hesitate to ping me if you have any doubts about it or if you need any further clarifications about move semantics. We can go over it again if you want.

The rest of the PR looks good, I'm confident once we address move semantics we can merge this.

Commit:
add Semantics
modifed const int row in matrix

glpuga commented on course/src/Vector3.cc:
```diff
@@ -4,43 +4,53 @@ namespace ekumen
+    Vector3::Vector3(Vector3&& other) {
+        values = other.values;
+        other.values = nullptr;
+    }
```
Good work !

glpuga commented on course/src/Vector3.cc:
```diff
@@ -75,13 +85,26 @@ namespace ekumen
+            values[0] = other.x(); 
+            values[1] = other.y();
+            values[2] = other.z();
+            other.values = nullptr;
```
Almost! There's a problem here. This is what this could should do:
```
            delete[] values;
            values = other.values;
            other.values = nullptr;
```
Notice I transfer the resource and empty the old vector. The way it's being done above it's leaking the memory that was allocated in `other.values`.

patapufate replied:
fixed

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,61 @@
+        Matrix3(const std::initializer_list<double> &list);
+        Matrix3(const std::initializer_list<double> &r1, const std::initializer_list<double> &r2, const std::initializer_list<double> &r3);
+        Matrix3(const Matrix3 &m);
+
```
Add a move constructor. Since your vectors already implement movement, it can be as simple as:
```
Matrix3(Matrix3 &&m) {
  row_values[0] = std::move(m.row_values[0]);
  row_values[1] = std::move(m.row_values[1]);
  row_values[2] = std::move(m.row_values[2]);
}
```

patapufate replied:
Done

glpuga commented on course/include/Matrix3.h:
```diff
@@ -0,0 +1,61 @@
+        Matrix3 &operator-=(const Matrix3 &q);
+
+        //operator overload
+        const Vector3 &operator[](const int &index) const;
```
Please also overload `operator=()`, both for copy and move.

patapufate replied:
Done

glpuga commented on course/src/Vector3.cc:
```diff
@@ -0,0 +1,211 @@
+
+    Vector3::Vector3(const Vector3 &p2)
+    {   
+        values = new double[3];
```
Remember to free the memory that was previously pointed to by `values`, or you'll leak memory!

patapufate replied:
it was fixed in previous PR.
    Vector3::~Vector3(){
        delete[] values;
    }

glpuga commented on course/test/src/isometry_TEST.cc:
```diff
@@ -125,27 +169,32 @@ GTEST_TEST(IsometryTest, IsometryOperations) {
+}   // namespace test
+}     // namespace cppcourse
+}       // namespace math
+} // namespace ekumen
```
nit: please align.

patapufate replied:
fixed

glpuga reviewed (approved):
Great work! Just a final round of comments, but it's approved. Fix and merge!

Commit:
Isometry Object fourth Review
